### PR TITLE
Setup and scrape tasks which define a set of scrapers and relevant scraping options

### LINKED ISSUE #28: Scraping multiple accounts together
Hello @eshaham. As a nYnaber for a year and a half I was using Finanda import module to scrape all the sites. As they decided to end the life of their desktop product I was looking for alternatives and found your two awesome repositories.

After playing with this library I think adding the following will contribute to its' daily flow:
1. extend `npm run credentials` to manage multiple account together (allowing the user to view list of configured accounts, modify/delete existing, add new ones) and name them as you wish, ie 'home accounts' or 'work accounts'
2. scrape a group of accounts and export a joined csv file to be easily import them together to nYnab.
3. change the layout of the exported csv to match the fields recommended by nYnab: account, date, payee, inflow, outflow, memo.

I think having supporting multiple accounts will reduce the number of key strokes significantly.

I would appreciate your insights :)

### PR DESCRIPTION
**Pull request by esakal**
### Motivation
This PR was created following #28. 

### A task structure
A task defines a set of scrapers and relevant scraping options. A user can define multiple tasks that serve different budgets (for example separated work and home budget). 

An example for a real task can be:
```
{
  "scrapers": [
    {
      "id": "leumi",
      "credentials": {
        "username": "",
        "password": ""
      }
    },
    {
      "id": "leumiCard",
      "credentials": {
        "username": "",
        "password": ""
      }
    }
  ],
  "options": {
    "combineInstallments": false,
    "dateDiffByMonth": 3
  },
  "output": {
    "saveLocation": "/Users/eransakal/Downloads/Transactions",
    "combineReport": true
  }
}
```

### Setup a  task
A user can manage tasks using the existing setup command `npm run setup`. I was working on the setup process to be as friendly as possible.

You can see the task setup menu structure [here](https://gist.github.com/esakal/b0556b3d0ade7a0ce352a30e73493e9a).

### New dependencies
- [colors](https://www.npmjs.com/package/colors) library was added to allow coloring feedback provided to the user while he/she setup a task. It is enriching his user experience.

### Linting adjustments
- **Ignoring no-underscore-dangle**: Since the setup process is quite complex, I split the code into multiple classes so it would be easier to read and maintain the code. The out-of-the-box rules of airbnb prevent using the convention _ to indicate a method/property should not be used as public api. I don't want to argue whether it is good or not (airbnb/javascript#1089 and airbnb/javascript#1024) but until ECMA classes will provide a decent support for private methods/properties I do think this is better then not distinguishing between private and public api. 
- **Ignoring class-methods-use-this**: This one is quite annoying because it prevent you from having a public api of a class which doesn't use other inner properties/methods of the class. It forces you setting that function as static which is very intrusive and limiting the programmer freewill.

### Splitting individual scrape code
A task is basically a group of individual scrapers with predefined scraping options. So basically we want to run the individual scraper multiple times when executing a task.
To keep DRY principle I needed to split the original [scrape-individual.js](https://github.com/eshaham/israeli-ynab-updater/blob/master/src/scrape/scrape-individual.js) file into two files. The first file handles the actual scraping and is used both by the individual scraper and the task scraper. The second file is the entry point when the user wants to scrape individual scraper and is used to get the relevant parameters from the user.

### Extracting the report  generation 
A task scraper support two types of output, a single report that combine all the scrapped accounts together and for backward compatible it also allow creating multiple reports, one per each account. 
For the same reason as explained above to keep the DRY principle I extracted the report generation into a separated file.

That's it, all the remaining files are there mostly to setup and scrape a task.

@eshaham Please review. Thanks!

Closes #28

---
This change is [Reviewable](https://reviewable.io/reviews/eshaham/israeli-ynab-updater/32)

### CODE REVIEW

File: src/setup/tasks/modify-task-handler.js [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+
+  constructor(taskName) {
+    this._taskName = taskName;
+  }
+
+  _getListOfScrapers() {
+    return Object.keys(SCRAPERS).map((scraperId) => {
+      const result = { value: scraperId, name: SCRAPERS[scraperId].name };
+      const hasCredentials = this._taskData.scrapers
+        .findIndex(scraper => scraper.id === scraperId) !== -1;
```
Comment (eshaham) [inline src/setup/tasks/modify-task-handler.js]: @esakal you can just use `find` here

File: src/setup/tasks/modify-task-handler.js [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+        message: 'Combine installment transactions?',
+        default: combineInstallments,
+      },
+      {
+        type: 'input',
+        name: 'dateDiffByMonth',
+        message: 'How many months do you want to scrape (1-12)?',
+        default: dateDiffByMonth,
+        filter: (value) => {
+          if (Number.isFinite(value) && !Number.isNaN(value)) {
```
Comment (eshaham) [inline src/setup/tasks/modify-task-handler.js]: @esakal does this ever happen? Doesn't `input` field type always returns a string?

File: src/setup/tasks/modify-task-handler.js [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+            return value;
+          } else if (typeof value === 'string' && value.match(/^[0-9]+$/)) {
```
Comment (eshaham) [inline src/setup/tasks/modify-task-handler.js]: @esakal wouldn't this be simpler?
```node
try {
  return parseInt(value, 10);
} catch (e) {
  console.log(...);
  return null;
}
```

File: src/setup/tasks/modify-task-handler.js [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+          },
+          goBackOption,
+        ],
+      },
+      {
+        type: 'list',
+        name: 'scraperId',
+        message: 'Select a scraper',
+        when: (answers) => {
+          if (answers.action === 'delete') {
```
Comment (eshaham) [inline src/setup/tasks/modify-task-handler.js]: @esakal would be nice you could extract the action names into an enum

**Comment by esakal** [comment]
@eshaham a spoiler alert, this is going to be a bit long comment, I tried to cover everything. I will work on the inline comments soon.

> I must say I can’t wait to be able to start using what you implemented here 👍

Thanks!  it feels very powerful to scrape all the accounts at the same time. 

> 1. The underscore notation of private fields makes me tick… 😄 . I would rather see functions extracted outside of the class if they are not to be shared publicly. At some point I gave up and stopped adding comments about it, but I would love to see it changed.

When I created the PR I was really hoping that you and I are on the same side of this debate :) I don't understand how private properties/methods were left out of the specification. It is be very hard for me to remove the `_` prefix, mostly because not having a way to let other developers distinguish between private and public api is like not having an api at all. I can live with moving private functions from the class even if it breaks the class integrity and makes the code less readable, but not having private properties is... well... But I respect your thought and opinion and will make the relevant adjustments to ditch the `_` prefix.

> This was a massive PR, and I really had a hard time reviewing all of it at once. I would prefer seeing smaller incremental PRs, but anyway… Given this comment, I never tend to re-review PRs, but in this case I reserve the right to review again 😄

We are on the same page, having smaller incremental PRs are easier to follow. This is the reason I created #29 and #30 to begin with. But I don't think there is an intuitive way to split the rest of the code. This PR deals with managing & scraping tasks and splitting those two tasks is irrelevant. As a side note, I would like to recommend an awesome reviewing tool that I integrated into my company development process which was essential to the success of our project and is called [reviewable.io](https://beta.reviewable.io/). it is marked as beta but I think it is like that just because they forgot to remove the label. I love this tool and I think it has everything you dream for in reviewing service. Obviously you don't need to use it with the project unless you want to give it a try, but I owe it to the developers of this service to talk about it whenever I can. 

> Would be great to see some reflection of this in the README file

Sure, will do

> @esakal you might want to close this issue using one of the relevant keywords in the PR description.

Done

> Thanks so much for all the effort you put into this!

It is the minimum I can do, I really appreciate your work on those libraries and your contribution to the open banking concept in Israel.

**Comment by eshaham** [comment]
> As a side note, I would like to recommend an awesome reviewing tool that I integrated into my company development process which was essential to the success of our project and is called reviewable.io

Could you TL;DR about what's so great about reviewable.io?
Took a peek at their website, but their explanation isn't that clear. I'm not sure what I would gain from their tool, which doesn't already exist on GitHub.

**Comment by esakal** [comment]
> Could you TL;DR about what's so great about reviewable.io?

I think the [following article](http://elliot.land/post/using-reviewable-io-for-code-review) layout nicely the reasons why reviewable.io transcends github reviewing support. I think it sums up to whether you believe that code reviews are essential part of the development process, if you do then once you will start using it you will not look back.

File: .eslintrc [hunk 5]
```diff
@@ -3,6 +3,8 @@
     "arrow-body-style": 0,
     "no-shadow": 0,
     "no-console": 0,
+    "no-underscore-dangle": ["error", { "allow": ["_private"] }],
```
Comment (esakal) [inline .eslintrc]: Ok, I found a compromise imo. 
- **for private methods** - I moved them outside of the class if they didn't use the state of the instance , for others I removed the prefix and set `@private` flag of jsdoc
- **for private properties** - I used the [following strategy](https://chrisrng.svbtle.com/using-weakmap-for-private-properties) to provide true support for private properties. I remember that you prefer not disabling lint by row so I added the exception of `_private` that as you can see is private by design.

**Comment by esakal** [comment]
@eshaham if you still find it difficult to review, I can split this one into two separated branches/prs:
1. the code that deals with scraping a task.
2. the code that deals with setup a task.

There are some complications with this separation:
1. You will lose the connection to the comments you already wrote
2. You will have a blocking dependency between them because there is no point adding one without the other. 
3. There are some shared additions/changes that will be duplicated between them and will cause merge conflicts.

I know it is hard to follow big reviews using github PR dashboard so if you think it worth the effort we can do it and tackle the complications as we go.

Please lmk if you want to continue with the PR or with the suggested alternative

**Comment by eshaham** [comment]
@esakal let's keep this PR, as we've already made some progress with it.
I've looked at reviewable.io, but didn't like the fact that mixing reviews in GH and reviewable will cause line comments to become top level comments, see [here](https://github.com/Reviewable/Reviewable/wiki/FAQ#can-i-mix-using-github-and-reviewable-on-a-pr).
I think I will stick to GH for now. 
Will go over your comments and changes and provide feedback soon.

**Comment by esakal** [comment]
thanks, I will wait for your feedback. Regarding reviewable.io, I also almost ruled out reviewable. But since my work project rely heavily on reviews we learned over time that once you start working with reviewable you don't feel the need to use github dashboard anymore. This is my personal experience but I understand your concerns about the comments becoming top level comments instead of inline comments.

File: src/scrape/generate-reports.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+  console.log(colors.notify(`${numFiles} csv files saved under ${saveLocation}`));
+}
+
+export async function generateSingleReport(scrapedAccounts, saveLocation) {
+  const fileTransactions = scrapedAccounts.reduce((acc, account) => {
+    acc.push(...account.txns);
+    return acc;
+  }, []);
+  const filePath = `${saveLocation}/${moment().format('DD-MM-YYYY_HH-mm-ss')}.csv`;
+  const fileFields = ['Institude', 'Account', 'Date', 'Payee', 'Inflow', 'Installment', 'Total'];
```
[the remark this answers is not in the document]
Comment (eshaham) [reply src/scrape/generate-reports.js]: > Using Bank here instead will not fit since we scrape also credit cards companies, can you recommend other alternatives?

hmm, ok. But I would love to be consistent. I've used "bank" in the past to describe a credit card account as well. I've also used `companyId`. Institute is an option I guess, but we will need to fix across both repos.
Your call.

> It is needed in single report since you need to associate each transaction to its scraper/account. The separated report doesn't need institute and Account but I don't mind adding them there as well for readability.

```node
const FIELDS = ['Date', 'Payee', 'Inflow', 'Installment', 'Total'];
const EXTRA_FIELDS = ['Institude', 'Account', ...FIELDS];
```
Wouldn't that be simpler?

File: src/scrape/generate-reports.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  const fileContent = json2csv({ data: fileTransactions, fileFields, withBOM: true });
+  await writeFile(filePath, fileContent);
```
[the remark this answers is not in the document]
Comment (eshaham) [reply src/scrape/generate-reports.js]: @esakal I meant something extracting code to a different function in this class. But now I see it may be more complicated and the benefit is too small, so let's drop it.

File: src/scrape/scrape-task.js [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+      }
+    }
+
+    console.log(colors.title('Save generated report'));
+
+    if (combineReport) {
+      const saveLocation = `${saveLocationRootPath}/tasks/${taskName}`;
+      await generateSingleReport(scrapedAccounts, saveLocation);
+    } else {
+      const currentExecutionFolder = moment().format('DD-MM-YYYY_HH-mm-ss');
```
[the remark this answers is not in the document]
Comment (eshaham) [reply src/scrape/scrape-task.js]: @esakal yeah, I got mixed emotions about this rule as well. Let's keep it for now though.

File: .eslintrc (hunk 5 above)
Comment (eshaham) [reply .eslintrc]: @esakal my way of saying sababa

File: src/helpers/tasks-manager.js [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+      this._writeSummaryLine('Combine installments', combineInstallments ? 'yes' : 'no');
+      this._writeSummaryLine('Save to location', saveLocation);
+      this._writeSummaryLine('Create single report', combineReport ? 'yes' : 'no');
+    }
+  }
+}
+
+const taskManager = new TaskManager();
+
+export default taskManager;
```
[the remark this answers is not in the document]
Comment (eshaham) [reply src/helpers/tasks-manager.js]: @esakal 😄 
"mizgun" comes out of the Hebrew saying, "Preparing for Air condition".
The meaning for it here is, setting up stuff in preparation for a future addition, which may or may not ever come.

[the remark this answers is not in the document]
Comment (esakal) [reply src/helpers/tasks-manager.js]: @eshaham It's funny I use this term a lot so I should have figure it out, I usually say הכנה למזגן so it is very close. :)

File: src/setup/tasks/modify-task-handler.js [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+                name: 'Update scraping options',
+                value: 'options',
+              },
+              goBackOption,
+            ],
+          },
+        ]);
+
+        switch (answers.action) {
+          case 'summary':
```
[the remark this answers is not in the document]
Comment (esakal) [reply src/setup/tasks/modify-task-handler.js]: done

**Comment by esakal** [comment]
@eshaham I think I covered all comments we currently have in this PR. 
- I changed file `tasks-manager.js` into `tasks.js` and separated the summary creation from the task manager. I also exported the class instead of a singleton.
- I'm sure you are familiar with this option but just in case, since we already have 17 commits, if you want you can always squash them into one single commit.

Night!

**Comment by eshaham** [comment]
@esakal I don't believe in squashing commits, I like to see the evolution of the PR in the commit list.
